src/devices/Board/Board.cs

```diff
@@ -0,0 +1,407 @@
+                    }
+                    else
+                    {
+                        PinReservation reservation = reservations.FirstOrDefault(x => x.Owner == owner);
+                        if (reservation == null)
+                        {
+                            throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because you are not a valid owner.");
+                        }
+
+                        if (reservation.Usage != usage)
+                        {
+                            throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because you are not the owner or the usage is wrong. Class {reservation.Owner} has reserved the Pin for {reservation.Usage}");
+                        }
+
+                        reservations.Remove(reservation);
+                        if (reservations.Count == 0)
+                        {
+                            _pinReservations.Remove(logicalPin);
+                        }
+                    }
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because it is not reserved.");
+                }
+            }
+        }
+
+        public abstract PinUsage DetermineCurrentPinUsage(int pinNumber);
+
+        /// <summary>
+        /// Override this method if something special needs to be done to use the pin for the given device.
+        /// Many devices support multiple functions per Pin, but not at the same time, so that some kind of
+        /// multiplexer needs to be set accordingly.
+        /// </summary>
+        /// <param name="pinNumber">The logical pin number to use.</param>
+        /// <param name="usage">The intended usage</param>
+        protected virtual void ActivatePinMode(int pinNumber, PinUsage usage)
+        {
+        }
+
+        public virtual void Initialize()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(ToString());
+            }
+
+            _initialized = true;
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            _disposed = true;
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private sealed class PinReservation
+        {
+            public PinReservation(int pin, PinUsage usage, object owner)
+            {
+                Pin = pin;
+                Usage = usage;
+                Owner = owner;
+            }
+
+            public int Pin { get; }
+            public PinUsage Usage { get; }
+
+            /// <summary>
+            /// Component that owns the pin (used mainly for debugging)
+            /// </summary>
+            public object Owner { get; }
+        }
+
+        public virtual GpioController CreateGpioController(int[] pinAssignment = null)
```

Comment:
I'm not sure what pinAssignment would do here

Reply:
That is to allow creating a GpioController that handles only a set of pins. I added it to be consistent with the other CreateXYZ()-Methods, where you can also specify the pins to use. 

When we do the IGpioPin approach discussed elsewhere, this would be a duplicate.